### ISSUE TITLE
Complete system lockup, misc. other bugs

### ISSUE THREAD

**Issue by WebDust21** [OpenShot/openshot-qt#3040]
I downloaded OpenShot for the first time yesterday (October 11, 2019); first from the Software Center on my laptop (2.4.1), and then switching to the latest version via APPIMAGE download from the website (2.4.4.)

My laptop: HP Elitebook 850 G2 (with 128GB SSD + 1TB HDD)
OS: Ubuntu 18.04 x64 (with KDE)
Processor: Intel Core i5-5200U @ 2.2GHz
RAM: 8GB.
(for the record, I did turn off "anonymous metrics", as I don't like the feeling of being observed by someone I don't know.  Turned them back on this morning for a second attempt at rendering the video.  Oh, and I've been writing this while waiting for the rendering...and I've also saved this post multiple times in GEdit.)

It's not a "speed demon" by any metric; while OpenShot glitches it's way through video playback (and really struggles with transitions), I'm not too surprised, as my laptop can just barely keep up with a 1080p:60 video stream.  No complaint there.

Up front: I have NEVER had OpenShot crash.  That's a plus.

My project: a 24-minute video comprising mostly of 1080p:60 M2TS (MTS) clips taken with my Sony DSC-HX100V digital camera (stereo sound).  There are a few transitions, a number of clip fades, a number of titles, and several still pictures.

First problem: while it does glitch it's way through playback (even with just one video on the timeline, no overlays or anything), that's probably attributable to my laptop.  What I consider to be a problem is when the video output section of OpenShot randomly locks up for 10-30 seconds at a time (most often around video transitions, or times there are several video lines to render).  I can still click things around the rest of the program--just I can't get it to play anything for up to 30 seconds.  If I play the transition again, it usually works without locking up (still very glitchy, though.)

MAJOR PROBLEM: Total system lockup.  I have had this happen TWICE, which is two times too many.  Both times, I had to terminate my laptop (hold the Power button down for 3 seconds, and then restart it)
1st time: I had been doing a lot of video editing, pulling clips together, adding transitions, etc.  It had been well over an hour of work.  I had had the video output section freeze several times by this point--definitely annoying, but nothing I couldn't live with.  But then one time I tried to play the video, instead of just the video output section locking up, OpenShot locked up my ENTIRE laptop.  The clock on my taskbar displays seconds; those weren't counting.  If I moved my mouse, the pointer on the screen would start clumsily jerking across the screen about five seconds later, and ten times slower.  Obviously, it was totally futile to try to recover the system; after waiting an awkward couple of minutes, I ended up force-restarting my laptop.

Upon restarting my laptop, I re-opened OpenShot (as the project wasn't done), and it promptly informed me (something along the lines of) that the backup of my project was successful, and did I want to open it?  Wow, such a nice program--sure!  
Guess what: it re-opened a project from the draft plans of the Stone Age, with just one video on the timeline.  I know that autosave was set to the default 3 minutes, but I also know I don't work THAT fast.  Re-opening the project file brought me back to right where I'd left off--what a relief.

Finishing the video, I started OpenShot rendering the output video (converting 1080p:60 M2TS files to a 720p:30 MP4).  The total estimated time was around 90 minutes @ ~8fps.  Oh, well, guess I'll do something else...
At 96.3%, OpenShot terminally locked up my laptop again.  I'd been listening to a video on YouTube in FireFox, and it started skipping and blipping like a badly scratched CD.  Upon closer investigation, I found that my mouse was back to the speed of frozen molasses--it was another terminal situation.  Obviously, after terminating and restarting my laptop, the 2.7GB resulting MP4 file was completely useless.  Worse, all the MP4 repair tools that I could find all required a "good video" to compare the "bad video" to--in short, it was a total loss.  So much for getting that project progress video out to my boss...

While I don't know for sure, I am suspecting that the total system lockup has something to do with "total time used" on the program, somewhere over 1.5 hours.

This morning when doing some experiments before attempting rendering again, I tried MP4 (H.265) video output; it pretty well completely locked OpenShot up (no progress updates, etc., though the output file was still growing).  Fortunately, it didn't terminally hang my laptop; I was able to terminate OpenShot and try again with MP4 (H.264) video output.

In addition: OpenShot v2.4.4 downloaded directly from the website has a minor GUI startup bug that wasn't present in v2.4.1.  This may in part be due to my dual-monitor setup, where the laptop's built in 1920x1080 screen is the primary, and I have a 1680x1050 Acer AL2216W as the secondary screen, rotated sideways, and configured to the right of the main laptop screen (it's physically above the laptop screen.)  I never have used the second screen with OpenShot.
![Displays](https://user-images.githubusercontent.com/13712655/66702993-98610980-ecdb-11e9-885e-7421064d5a60.png)

Anyway, when opening OpenShot, it appears as I would expect.  When I maximize it, well, it doesn't quite look like I expected it should (timeline is all squished, with a LOT of wasted space beneath it):
![OpenShot Resize Bug](https://user-images.githubusercontent.com/13712655/66703158-84b6a280-ecdd-11e9-93ea-8699e173cb30.png)

P.S. Re-rendering the video at a lower quality produced a 941MB output file, took just over an hour, and successfully completed rendering.  FWIW I also turned off "Process Video Frames In Parallel (Experimental)" in the settings.  Seemed to make video playback a lot more glitchy, but at least it didn't terminally hang my laptop.

**Comment by ferdnyc** [comment]
@WebDust21  Thanks for the detailed feedback! Glad to hear you were _mostly_ successful working with OpenShot, though I realize your experience wasn't without some stumbles and frustrations along the way. 

> (for the record, I did turn off "anonymous metrics", as I don't like the feeling of being observed by someone I don't know.

And you're absolutely more than welcome to! There's a reason that prompt is literally the _very first_ interaction new users have with OpenShot, nobody should feel pressured into providing data collection they're not comfortable with.

(The data collection is also performed through Google Analytics, which has proved to provoke a knee-jerk negative — if not downright angry — reaction from some people. All I can say is that unlike a web browser or etc., OpenShot doesn't _have_ any personally identifying information about its users. It has no access to tracking cookies, or browsing history, or whatever else. Literally the only information sent in the metrics is "At this time, a user associated with a particular `unique_install` identifier — a random GUID not connected to any other identifying information — ran OpenShot version X.Y.Z (with these versions of libopenshot, Python, Qt, PyQt5), opened X dialog, performed a Y action, exported a video with these parameters, encountered a crash in code module A that produced traceback B..." Literally every person who posts an issue here on GitHub ends up revealing _way_ more detail about themselves for Google or anyone else to collect.)

> First problem: while it does glitch it's way through playback (even with just one video on the timeline, no overlays or anything), that's probably attributable to my laptop.

It's kind of not, honestly. Preview playback is... poor, still, and has thus far proved annoyingly resistant to optimization. While it may be true that the performance you experienced is to be expected on your system, unfortunately users with much higher-end hardware don't really fare _that_ much better. It's an ongoing pain point.

> What I consider to be a problem is when the video output section of OpenShot randomly locks up for 10-30 seconds at a time (most often around video transitions, or times there are several video lines to render). I can still click things around the rest of the program--just I can't get it to play anything for up to 30 seconds. If I play the transition again, it usually works without locking up (still very glitchy, though.)
> 
> MAJOR PROBLEM: Total system lockup. I have had this happen TWICE, which is two times too many. Both times, I had to terminate my laptop (hold the Power button down for 3 seconds, and then restart it)

That's almost certainly a memory-exhaustion issue. Most likely your system's RAM became vastly overcommitted, and you ended up in what, for the past 20-30 years now, has been not-so-affectionately known as "swap hell". There are a lot of pieces that go into creating these situations:

1. 8GB of RAM is a little tight for HD video editing, honestly. Especially at 60fps, and _especially_-especially when the GPU doesn't have any dedicated memory of its own, and is going to have to share that system RAM.

2. OpenShot's memory management is... less-than-ideal, in many situations.

   In 2.4.4 release, specifically, there's a rather severe issue with excessive memory consumption during the export process — the more individual, separate _elements_ your project timeline has, the more memory the export process will consume as it progresses, because the memory allocated for each input stream wasn't being properly released when it was no longer needed. Nearly all of it stayed allocated until the completion of the entire Export, at which point it would all finally be freed at once. (If you managed to make it there.)

   Fortunately, that was fixed in OpenShot/libopenshot#240 not too long after 2.4.4 was completed, but since there hasn't been a release since (and it's still going to be a while longer before the next one, realistically), this will probably not be the last time I suggest checking out the latest Daily Build releases (there's a button on the same https://openshot.org/download/ page), because you _may_ find that the latest code improves on some of the problems you encountered.

3. Linux handles these situations _very, **very**_ poorly. Still. Somehow. Even though it's been a known problem for, like I said, more than 2 _decades_ already, progress on it has been extremely slow. Even though everyone knows it's a problem, everyone agrees that it's a problem, and everyone is on board with the idea that it really needs to be dealt with.

   The only encouraging information I can offer here is that there's been _some_ small improvement. I definitely experience the same sort of issues — not often with OpenShot, honestly (I do very little video editing, mostly just light testing/debugging), usually for me it's a web browser with too many tabs open. Or I've got like a 50/50 shot if I fire up my Windows 7 VM, since that grabs a full 3GB of RAM for itself. But my main desktop system is extremely similar to yours (4-core Core i5 CPU, 8GB RAM, consumer-level SSD; the one advantage I have is a PCIe Nvidia GTX 710 GPU, which even though it's an underpowered $50 cheapie does at least bring an additional full gig of dedicated video RAM to the party), and it can definitely be a tightrope act.

   _But_, that being said (I did say this was positive), only **very** recently (since earlier this year) I've noticed some changes in the nature of those swap-hell lockups, the ones where the system just tips over into an infinite well of paging activity where it will never find the bottom. At least on my Fedora 30 machine, what I've found is that switching to a text console using <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd><var>number</var></kbd> seems to pull the kernel virtual memory system out of its own ass and regain control. It may still take 15-30 seconds for the login prompt to come up, but when it does I can usually log in within a reasonable timeframe. At that point I can terminate the runaway memory hog, if necessary — but often enough I'll find that the `oom-killer` has already taken care of it.

   The switch to a different virtual console seems to have gained new powers to preempt the paging system and get it to settle down, instead of having to wait for it to complete _all_ of its queued activity — which will typically take many, many minutes, if indeed it's ever going to finish at all. It's still a too-manual process, and if I _don't_ take action then the system will happily sit there freaking itself out for 10-15 minutes minimum (or until I lean on the power button, because nobody's that patient), but at least it's _some_ improvement. I'm guessing that's down to changes in recent kernels or newer systemd, so i could't say whether it would apply to Xenial systems. But, maybe something to look forward to in a future release, if nothing else?

As far as OpenShot's memory issues go, I guess I have two primary pieces of advice:

1. Don't export at High quality. For H.264 MP4 and many other codecs, it sets a fixed video bitrate of something like 15 Mbps, which is _severe_ overkill. (But also matches YouTube's recommended bitrate for full-HD uploads.) Unless you actually need that high a bitrate, Medium quality is still 5 Mbps which is typically more than adequate. (And when it far exceeds the _input_ video bandwidth, is also really pointless **unless** you're using a lot of effects or adding rendered elements that can mean the output quality _isn't_ completely limited by the source quality.

2. Because OpenShot is very sloppy with RAM allocations and can really fragment the system over time, it's often a good idea to precede an Export operation by saving the final version of the project, quitting OpenShot completely, then reloading the project and heading _immediately_ for the Export dialog. Most especially, don't Preview anything prior to exporting. That'll help make as much memory as possible available for use by the Export process.

> Guess what: it re-opened a project from the draft plans of the Stone Age, with just one video on the timeline. I know that autosave was set to the default 3 minutes, but I also know I don't work THAT fast. Re-opening the project file brought me back to right where I'd left off--what a relief.

Hmmm. Sounds like we might be failing to _remove_ our backup files, after we successfully write a more up-to-date autosave to the project save location. Backups are meant to protect against the loss of changes only _since_ the last save, which means a backup file older than the save data is pretty pointless. I'll take a look and see if something's not being updated, or kept around beyond its usefulness.

> Anyway, when opening OpenShot, it appears as I would expect. When I maximize it, well, it doesn't quite look like I expected it should (timeline is all squished, with a LOT of wasted space beneath it)

Yeah, the default layout in 2.4.4 is just nonsense. See #2307. 

Window layout is obviously a subjective thing, and everyone will have their own opinions on that issue. You certainly can't ever hope to satisfy everyone. But, still, there are reasonable baselines that most people can agree on, and that nonsense is none of them. As usual.

Every single OpenShot release, seems like, and typically _**just** before_ the actual release, a commit is pushed into the repo that updates the stored default layouts in some comically impractical way. I don't know _why_, I have no **IDEA** what's driving those decisions and what goals they're attempting to achieve, but at this point I can't help but feel like it's being done specifically to torture me, personally.

Anyway, OpenShot 2.4.4 managed to reach new heights in impracticality with the latest default Advanced view, and I have PR #2987 open to correct that. If past history is anything to go on, most likely that will get merged at some point, with everyone agreeing it seems fine, leading me to assume the issue is settled. And then, right before the release announcement is made for version 2.5, close enough that I won't even discover it until _after_ the release is already out the door, a commit will get checked in that turns the entire interface inside-out, or some other insane bullshit. At this point it's just tradition, and feels as inevitable as the tides.

I guess all I can tell you is — despite defaults that look like what you'd end up with if someone's grandmother sat down at the computer and tried to use OpenShot to see her grandkids' graduation photos "in The Google" — the interface is actually more flexible than it's ever been, and with a little rearranging you can get it to a state where it's pretty comfortable regardless how you work or how much screen real estate you have. 

(In your case, for instance, you COULD easily make good use of that second monitor — both the Preview dock and the Timeline dock are detachable from the main window, and either one could make good use of the second-screen real estate, and free up a lot of room for the rest of the UI to spread out more.)

Just remember not to ever activate "Simple View" or "Advanced View" after that, because they both force a _complete_ reset of the entire GUI back to the corresponding stored configuration. All of your customizations will be wiped out, and you'll have no option but to manually do them all over again. Until then, you'll be stuck in the useless insanity that is the standard default layout, because welcome to my hell.

> FWIW I also turned off "Process Video Frames In Parallel (Experimental)" in the settings. Seemed to make video playback a lot more glitchy, but at least it didn't terminally hang my laptop.

There again is an area where the Daily Builds are worth exploring — in addition to better behind-the-scenes resource management, the Performance tab of Preferences has been _greatly_ expanded with new tuning options that give you more control over OpenShot's resource consumption, in particular during export. So now it's possible to _reduce_ parallelism (and therefore memory consumption, because the memory allocations of the export process are multiplied by the number of threads in use), rather than it being an all-or-nothing sort of thing.

![image](https://user-images.githubusercontent.com/538020/66718037-4ee0ef00-edad-11e9-8e0a-9a2bac2a98d4.png)

**Comment by ferdnyc** [comment]
Whoops, wrong image! I meant this one.

![Screenshot from 2019-10-13 11-32-09](https://user-images.githubusercontent.com/538020/66718100-037b1080-edae-11e9-8124-6c2939d94767.png)

**Comment by WebDust21** [comment]
Thanks for the detailed response ;-).  

I wasn't complaining about the window layout, as much as a bug with window layout.  Here's how OpenShot looks if I just "restore" the window to a non-maximized state:
![Restored](https://user-images.githubusercontent.com/13712655/66724010-75734a00-edee-11e9-93e7-1de723bfa2d8.png)
All of the wasted space beneath the timeline is gone.  I will also note that when maximized, I can't simply drag the lower "adjustment" down...not without moving the top one down, then moving the bottom one down (which doesn't make any sense).
And it looks fine "restored", but if I maximize it, OpenShot goes right back to the other view, with a large empty space beneath the timeline.  
(Seems that the bug originates in window resizing: if I resize OpenShot smaller, everything resizes as expected.  But if I resize OpenShot taller, the timeline does not move back down--it stays put wherever the smallest state was.)

> (The data collection is also performed through Google Analytics, which has proved to provoke a knee-jerk negative — if not downright angry — reaction from some people. **All I can say is that unlike a web browser or etc., OpenShot doesn't have any personally identifying information about its users.** 
I didn't hear anything you wrote after the word, "Google" :-).

> As far as OpenShot's memory issues go, I guess I have two primary pieces of advice:
> 
>     Don't export at High quality. For H.264 MP4 and many other codecs, it sets a fixed video bitrate of something like 15 Mbps, which is severe overkill. (But also matches YouTube's recommended bitrate for full-HD uploads.) Unless you actually need that high a bitrate, Medium quality is still 5 Mbps which is typically more than adequate. (And when it far exceeds the input video bandwidth, is also really pointless unless you're using a lot of effects or adding rendered elements that can mean the output quality isn't completely limited by the source quality.

That's what I ended up doing (medium quality), and it worked.  Didn't need the high quality; they were low-light 28Mbps M2TS videos anyway.

>     Because OpenShot is very sloppy with RAM allocations and can really fragment the system over time, it's often a good idea to precede an Export operation by saving the final version of the project, quitting OpenShot completely, then reloading the project and heading immediately for the Export dialog. Most especially, don't Preview anything prior to exporting. That'll help make as much memory as possible available for use by the Export process.

I'd begun to guess as much...that's what I did on the final rendering: fixed up the project (broken due to another bug--this one I can pinpoint identify), restarted OpenShot, started the rendering project, and crossed my fingers.

The one last bug, which caused major frustration when I tried to re-encode the project the first time (before just giving up and going to bed) has to do with the Titles (Ctrl+T), and filenames.
Basically, when you open the Titles window, it gives a default filename to the title; in my case, it's "TitleFileName-1".  The next title will be "TitleFileName-2", and so on.  (No problem there.)  When creating my video, I didn't bother to change the filenames; I just added the titles in sequence, and dragged them onto the timeline.
The problem started when OpenShot terminally locked up my laptop the first time.  After restarting my computer, I went right back to work on OpenShot...including adding two more titles to the video. As might be expected, since it had just been started, OpenShot re-started title numbering with "TitleFileName-1".  The problem: OpenShot didn't notify me that it was overwriting my previous title with the exact same filename.  I only discovered this problem after the failed render and subsequent lockup, when I went back to the beginning of the video...and found that the opening few titles had been overwritten with the new titles I had added to the end of the video.  Worse yet, the thumbnail picture in the "Project Files" for said titles, still showed the first title, and not what displayed if I placed the title on the timeline.

Thank you for the excellent product support; however, just to be frank and up front with you, I plan on trying a different video editing program in the future.  Terminal lockup is just not something I can recall experiencing before on my laptop--and I often have numerous large programs running.

**Comment by ferdnyc** [comment]
> > Obviously, it was totally futile to try to recover the system; after waiting an awkward couple of minutes, I ended up force-restarting my laptop.
> > 
> > Upon restarting my laptop, I re-opened OpenShot (as the project wasn't done), and it promptly informed me (something along the lines of) that the backup of my project was successful, and did I want to open it? Wow, such a nice program--sure!
> > 
> > Guess what: it re-opened a project from the draft plans of the Stone Age, with just one video on the timeline. I know that autosave was set to the default 3 minutes, but I also know I don't work THAT fast. Re-opening the project file brought me back to right where I'd left off--what a relief.
> > 
> 
> Hmmm. Sounds like we might be failing to remove our backup files, after we successfully write a more up-to-date autosave to the project save location. Backups are meant to protect against the loss of changes only since the last save, which means a backup file older than the save data is pretty pointless. I'll take a look and see if something's not being updated, or kept around beyond its usefulness.

For anyone reading this somewhere down the road, the reason this issue is open and labeled as a bug (since there's a lot going on here) is specifically this old-backup-files problem. (The other things discussed either already have open issues tracking them, or aren't really actionable bugs.)

**Comment by ferdnyc** [comment]
> I wasn't complaining about the window layout, as much as a bug with window layout.
> And it looks fine "restored", but if I maximize it, OpenShot goes right back to the other view, with a large empty space beneath the timeline.
> (Seems that the bug originates in window resizing: if I resize OpenShot smaller, everything resizes as expected. But if I resize OpenShot taller, the timeline does not move back down--it stays put wherever the smallest state was.)

The resizing bug, to the extent we have control over it, is caused by the layout. It all comes down to Qt's `QMainWindow` view, and specifically the fact that the Timeline is set as the `CentralWidget`, which Qt treats specially. Now that it's a dock widget, t's _supposed_ to be (I thought!) in the `BottomDockWidgetArea`, where it would hug the bottom of the window at all times like it's meant to.

> I didn't hear anything you wrote after the word, "Google" :-).

If you know another no-cost service for collecting anonymous telemetry from an unlimited number of users worldwide, I'm all ears.

If I had my way the collection would be opt-**IN** instead of opt-out, but I lost that argument and other than that, i really don't know any reasonable alternatives.

> The one last bug, which caused major frustration when I tried to re-encode the project the first time (before just giving up and going to bed) has to do with the Titles (Ctrl+T), and filenames.

Yeah, that's #2961, with a fix in progress at #2976.
